### PR TITLE
Acquisition dimension order in storage API

### LINKED ISSUE #31: Support configurable append dimension in StorageProperties
- [x] Modify the storage properties struct, storage property capabilities struct, and associated storage property configuration functions to allow configuration of
  - which dimensions above y and x will be used at all; and
  - which dimension should be the append dimension.
- [x] Test and, if necessary, modify non-Zarr storage devices to indicate that they do not support this functionality.

### PR DESCRIPTION
**Pull request by aliddell**
Allows users to specify the names, ordering, and number of acquisition dimensions.

### CODE REVIEW

File: acquire-core-libs/src/acquire-device-properties/device/props/storage.c [hunk 1]
```diff
@@ -82,30 +82,120 @@ storage_properties_set_external_metadata(struct StorageProperties* out,
...
 {
     CHECK(out);
-    out->chunk_dims_px.width = chunk_width;
-    out->chunk_dims_px.height = chunk_height;
-    out->chunk_dims_px.planes = chunk_planes;
+
+    EXPECT(name, "Dimension name cannot be null.");
+    EXPECT(bytes_of_name > 0, "Bytes of name must be positive.");
+    EXPECT(strlen(name) > 0, "Dimension name cannot be empty.");
+    EXPECT(kind > DimensionType_None && kind < DimensionTypeCount,
```
Comment (andy-sweet) [inline acquire-core-libs/src/acquire-device-properties/device/props/storage.c]: This prohibits explicitly using `DimensionType_None` with `dimension_init`, which was surprising to me. Let's say I have a named dimension that doesn't really have a type/kind - should I init it with `Channel`?

File: CHANGELOG.md [hunk 2]
```diff
@@ -15,6 +15,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Changed
 
 - `reserve_image_shape` is now called in `acquire_configure` rather than `acquire_start`.
+- Users can now specify the names, ordering, and number of acquisition dimensions, as well as the append dimension.
```
Comment (aliddell) [inline CHANGELOG.md]:
```suggestion
- Users can now specify the names, ordering, and number of acquisition dimensions.
```

File: acquire-core-libs/src/acquire-device-properties/device/props/storage.c (hunk 1 above)
Comment (aliddell) [reply acquire-core-libs/src/acquire-device-properties/device/props/storage.c]: Getting rid of `DimensionType_None`.

Comment (aliddell) [inline acquire-core-libs/src/acquire-device-properties/device/props/storage.c]:
```suggestion
    EXPECT(kind < DimensionTypeCount,

```

File: acquire-core-libs/src/acquire-device-properties/device/props/storage.c [hunk 3]
```diff
@@ -183,6 +273,33 @@ storage_properties_destroy(struct StorageProperties* self)
...
+               sizeof(self->acquisition_dimensions));
+    }
+}
+
+const char*
+dimension_type_as_string(enum DimensionType type)
+{
+    switch (type) {
+        case DimensionType_None:
+            return "None";
```
Comment (aliddell) [inline acquire-core-libs/src/acquire-device-properties/device/props/storage.c]:
```suggestion

```

File: acquire-core-libs/src/acquire-device-properties/device/props/storage.c [hunk 4]
```diff
@@ -368,20 +485,125 @@ unit_test__storage__copy_string()
...
+    // can't init with an empty name
+    CHECK(!dimension_init(&dim, "", 1, DimensionType_Spatial, 1, 1, 1));
+    CHECK(dim.name.str == NULL);
+    CHECK(dim.kind == DimensionType_None);
+    CHECK(dim.array_size_px == 0);
+    CHECK(dim.chunk_size_px == 0);
+    CHECK(dim.shard_size_chunks == 0);
+
+    // can't init with an invalid dimension type
+    CHECK(!dimension_init(&dim, "x", 2, DimensionType_None, 1, 1, 1));
```
Comment (aliddell) [inline acquire-core-libs/src/acquire-device-properties/device/props/storage.c]:
```suggestion
    CHECK(!dimension_init(&dim, "x", 2, DimensionTypeCount, 1, 1, 1));

```

File: acquire-core-libs/src/acquire-device-properties/device/props/storage.h [hunk 5]
```diff
@@ -11,6 +11,34 @@ extern "C"
     struct DeviceManager;
     struct VideoFrame;
 
+    enum DimensionType
+    {
+        DimensionType_None = 0,
+        DimensionType_Spatial,
```
Comment (aliddell) [inline acquire-core-libs/src/acquire-device-properties/device/props/storage.h]:
```suggestion
        DimensionType_Space = 0,

```

File: acquire-core-libs/src/acquire-device-properties/device/props/storage.h [hunk 6]
```diff
@@ -11,6 +11,34 @@ extern "C"
     struct DeviceManager;
     struct VideoFrame;
 
+    enum DimensionType
+    {
+        DimensionType_None = 0,
+        DimensionType_Spatial,
+        DimensionType_Channel,
+        DimensionType_Time,
+        DimensionTypeCount
```
Comment (aliddell) [inline acquire-core-libs/src/acquire-device-properties/device/props/storage.h]:
```suggestion
        DimensionType_Other,
        DimensionTypeCount

```

File: acquire-core-libs/src/acquire-device-properties/device/props/storage.h [hunk 7]
```diff
@@ -19,41 +47,39 @@ extern "C"
...
     {
-        /// Metadata for chunking.
-        /// Indicates whether chunking is supported, and if so, bounds on what
-        /// the dimensions (in px) of the chunks are.
         struct storage_property_metadata_chunking_s
         {
             uint8_t is_supported;
-            struct Property width, height, planes;
-        } chunk_dims_px;
+        } chunking;
```
Comment (nclack) [inline acquire-core-libs/src/acquire-device-properties/device/props/storage.h]: consider flattening these structs

File: acquire-core-libs/src/acquire-device-properties/device/props/storage.c [hunk 8]
```diff
@@ -183,6 +271,25 @@ storage_properties_destroy(struct StorageProperties* self)
             memset(strings[i], 0, sizeof(struct String)); // NOLINT
         }
     }
+
+    storage_properties_dimensions_destroy(self);
+}
+
+const char*
+dimension_type_as_string(enum DimensionType type)
```
Comment (nclack) [inline acquire-core-libs/src/acquire-device-properties/device/props/storage.c]: private?

**Review by nclack — APPROVED** [review]
some nits

**Pull request merged**